src/engine/renderer/tr_image.cpp

```diff
@@ -1057,6 +1431,58 @@ void R_UploadImage( const byte **dataArray, int numLayers, int numMips,
 		}
 	}
 
+	/* Clamp to the current upper OpenGL limit, scale both axis down equally so we don't have to
+	deal with a half mip resampling.
+
+	From Khronos OpenGL wiki:
+
+	> For 1D and 2D textures (and any texture types that use similar dimensionality,
+	> like cubemaps) the max size of either dimension is GL_MAX_TEXTURE_SIZE.
+	> For array textures, the maximum array length is GL_MAX_ARRAY_TEXTURE_LAYERS.
+	> For 3D textures, no dimension can be greater than GL_MAX_3D_TEXTURE_SIZE in size.
+	-- https://www.khronos.org/opengl/wiki/Texture
+
+	The OpenGL Red Book says:
+
+	> glGetIntegerv(GL_MAX_TEXTURE_SIZE,...) tells you the largest dimension (width or height,
+	> without borders) of a texture image, typically the size of the largest square texture
+	> supported. However, GL_MAX_TEXTURE_SIZE does not consider the effect of the internal
+	> format of a texture. A texture image that stores texels using the GL_RGBA16 internal
+	> format may be using 64 bits per texel, so its image may have to be 16 times smaller
+	> than an image with the GL_LUMINANCE4 internal format. (Also, images requiring borders
+	> or mipmaps may further reduce the amount of available memory.) 
+	-- http://www.glprogramming.com/red/chapter09.html
```

Comment:
How do you know which texture format GL_MAX_TEXTURE_SIZE is assuming?

It seems better to use the the method described in the next section, *Texture Proxy*. In that case, it is not necessary to have a database of the number of bits per texel, or to guess at what bit density the maximum value assumes, or whether one should consider the edge length or the total number of pixels.

Reply:
Things like `GL_LUMINANCE4` that is taken as example and others are not described in current OpenGL wiki, or I did not found them:
https://www.khronos.org/opengl/wiki/GLAPI/glTexImage2D

But we can find some of them in OpenGL 2.1 documentation:
https://www.khronos.org/registry/OpenGL-Refpages/gl2.1/xhtml/glTexImage2D.xml
Or in OpenGL 1.1 documentation:
https://web.archive.org/web/20011231003042/http://www.opengl.org/developers/documentation/Specs/glspec1.1/node79.html

It still don't find formal definition for `GL_RGB`.

I have read people saying GL_RGB is 24 bit, interlaced (`RGBRGBRGB`), like [here](https://forum.openframeworks.cc/t/gl-rgb/72/2). That's not from Khronos though, but that's an example of an answer to solve someone's need to load data directly as `GL_RGB`.

While using `imagelist` command I discovered some of our textures are listed as using `GL_RGBA`, so I may add those formats too:

```
      -w-- -h-- -px- -size------ -mm- -type- -format- -wrap.s- -wrap.t- -name------- 
Debug: Unknown pixel size for undocumented image format 6408 
Debug: Undocumented image format 6408 for image fonts/unifont_0_0_16.png 
  35:  512  256   32      174762 no   2D      6408    s.clmp   t.clmp   fonts/unifont_0_0_16.png 
```

`6408` is `0x1908` wich is `GL_RGBA` in `gl.h`…

So, unless proven wrong,it looks like `GL_RGB` uses 8 bit per pixel, then I see no reason why `GL_RGBA` would not use 8 bit per pixel, but documentation about them is harder to find. Also, unless proven wrong, it looks like `GL_RGB` is the default format.

Reply:
This code does not exist anymore.